Fix #2355: Shifted helper function to central testing utility

Akshat-gour opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR Fixes #2355: In which we Shifted helper function **hasErrorText** and **hasNoErrorText** from **org.oppia.android.app(test)** > profile > ( AddProfileActivityTest.kt , AdminAuthActivityTest.kt , AdminPinActivityTest.kt , PinPasswordActivityTest.kt )
To org.oppia.android.testing > **TextInputAction.kt** and added Context Parameter to TextInputAction Constructor and added Material Dependency to build.gradle(:testing) and to build.bazel(testing).
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Akshat-gour committed:
shifted helper function to central testing utility module

Akshat-gour committed:
did few change in format in files

Akshat-gour commented:
@anandwana001  this are two check which failed how can I solve then
![image](https://user-images.githubusercontent.com/60266468/107882162-d1337300-6f0d-11eb-9ef4-91a15b326c66.png)
![image](https://user-images.githubusercontent.com/60266468/107882213-0f309700-6f0e-11eb-87a1-6f476e26821e.png)

anandwana001 commented:
I had re-run the checks, checking if everything passes!!

Akshat-gour commented:
@anandwana001 
Even now 2 check Failed, anything which helps to pass the test

Akshat-gour commented:
@anandwana001 
please help me out with checks

anandwana001 commented:
Trying re-run, let's see if it works out.

Akshat-gour commented:
@anandwana001 
now also check are failing

anandwana001 commented:
@Akshat-gour Could you please are these test failing locally too or not?

anandwana001 committed:
Merge branch 'develop' into shifted-helper-function-to-central-testing-utility

Akshat-gour commented:
@anandwana001 
how can i run test locally

anandwana001 commented:
https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#running-robolectric-tests

Akshat-gour commented:
@anandwana001 
Do i have to run test only for files I edited like AdminAuthActivityTest ,  AddProfileActivityTest , AdminPinActivityTest ,
PinPasswordActivityTest .

And TextInputAction.kt does not have any test file

anandwana001 commented:
Actually, I just waiting one last time, If the tests passes on checks. Let's wait if it worked out.

Akshat-gour commented:
ok

anandwana001 commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,40 @@
+import org.hamcrest.TypeSafeMatcher
+import javax.inject.Inject
+
+class TextInputAction @Inject constructor(
```
Add `kdoc` to help developers understand what does this class do and when to use it. 
Refernce - `EditTextInputAction`

Akshat-gour replied:
@anandwana001 
Will it be correct to write

Action for accessing HasErrorTest and HasNoErrorTest Functions.

This is needed because HasErrorTest Function and HasNoErrorTest
Function can now be accessed from this file wherever needed
Without Implementing it.

Akshat-gour replied:
**or** reason is same as **EditTextInputAction** File
This is needed because Robolectric doesn't seem to properly input digits for text views using
'android:digits' or other filters. See `https://github.com/robolectric/robolectric/issues/5110`
for specifics.

Akshat-gour replied:
@anandwana001 
PTAL

anandwana001 replied:
Sounds good, need for robolectric

anandwana001 replied:
@Akshat-gour

Akshat-gour replied:
done

anandwana001 commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,40 @@
+class TextInputAction @Inject constructor(
+  val context: Context
+) {
+
```
Add todo `// TODO(#1720): Move this to a companion object & use a test-only singleton injector to retrieve`

Also, Please check this #1720 and let me know if we can fix this file in this PR only if not, no worries.

Akshat-gour replied:
@anandwana001 
I didn't understand this issue. I guess I need to implement companion object of TextInputAction class

Akshat-gour replied:
PTAL

anandwana001 replied:
I think we can skip this for now, this can be handle when we introduce dagger hilt.

anandwana001 commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,40 @@
+      }
+
+      override fun describeTo(description: Description) {
+        description.appendText("")
```
Could you please update this line with

```kotlin
description.appendText("TextInputLayout's error")
```

Akshat-gour replied:
done

anandwana001 replied:
Please say done after pushing the code, as for this pr this doesn't seems like done. Helpful to review.

anandwana001 replied:
@Akshat-gour

Akshat-gour replied:
done

anandwana001 reviewed (changes requested):
Thanks @Akshat-gour 

There are a couple of places left to shift like `ProfileRenameActivityTest`, `ProfileResetPinActivityTest`. 
Please look for any other left to use `TextInputAction`.

Akshat-gour committed:
Merge remote-tracking branch 'upstream/develop' into shifted-helper-function-to-central-testing-utility

Akshat-gour commented:
@anandwana001 
need help to resolve this type of conflict
![image](https://user-images.githubusercontent.com/60266468/108544460-7d110000-730c-11eb-8159-b1fcbe21204e.png)

Akshat-gour committed:
Removed helper function from ProfileRenameActivityTest.kt and ProfileResetPinActivityTest.kt and used
TextInputAction.kt file instead
Inside TextInputAction.kt added comments and ToDo

Akshat-gour committed:
removed used import and trailing spacing

Akshat-gour commented:
@anandwana001 
How to resolve this issue
![image](https://user-images.githubusercontent.com/60266468/108555923-32977f80-731c-11eb-867e-01185c9b47e9.png)

anandwana001 commented:
Keep the function name short, don't extend them over 100 char. Look for how to have a 100 char mark in Android Studio. (Grey line)

Akshat-gour committed:
reformatted the code

Akshat-gour committed:
reformatting the code

Akshat-gour committed:
reformatted the code

Akshat-gour committed:
Reformatted the code in ProfileResetPinActivityTest.kt file

Akshat-gour commented:
@anandwana001 
showing file changed after merging **upstream/develop** but i didn't changed this file 
![image](https://user-images.githubusercontent.com/60266468/108634945-5553a080-74a2-11eb-9bad-7ebe6dbbce87.png)

While trying to push the code showing this error

![image](https://user-images.githubusercontent.com/60266468/108634954-669cad00-74a2-11eb-860d-befc2d2558b4.png)

anandwana001 commented:
@FareesHussain could you help @Akshat-gour with hit commands? 

@Akshat-gour reference - https://www.atlassian.com/git/tutorials/syncing/git-pull#:~:text=The%20git%20pull%20command%20is,repository%20to%20match%20that%20content.&text=Once%20the%20content%20is%20downloaded,point%20at%20the%20new%20commit.

FareesHussain commented:
@Akshat-gour This is probably because this branch was updated using github. 
You need to run `git pull origin shifted-helper-function-to-central-testing-utility`

Akshat-gour committed:
Merge remote-tracking branch 'upstream/develop' into shifted-helper-function-to-central-testing-utility

# Conflicts:
#	app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt

Akshat-gour committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into shifted-helper-function-to-central-testing-utility

Akshat-gour committed:
Merge remote-tracking branch 'upstream/develop' into shifted-helper-function-to-central-testing-utility

Akshat-gour committed:
Merge branch 'shifted-helper-function-to-central-testing-utility' of https://github.com/Akshat-gour/oppia-android into shifted-helper-function-to-central-testing-utility

Akshat-gour commented:
@anandwana001 
PTAL
test are now also failing

FareesHussain commented:
> @anandwana001
> PTAL
> test are now also failing

The failing test is optional.
The tests with `Required` tag are necessary for the PR to get merged.

Akshat-gour commented:
but one with tag required is also failing

FareesHussain commented:
> but one with tag required is also failing

Just checked this can you try running the failing test case locally.

AdminAuthActivityTest > testAdminAuthActivity_defaultButtonState_isDisabled using robolectric.

Akshat-gour commented:
test fails locally also

FareesHussain commented on app/src/main/res/layout/admin_auth_activity.xml:
```diff
@@ -110,7 +110,7 @@
           android:layout_marginBottom="44dp"
           android:background="@{viewModel.isSubmitButtonActive ? @drawable/state_button_primary_background :@drawable/start_button_transparent_background}"
-          android:enabled="@{viewModel.isSubmitButtonActive}"
+          android:clickable="@{viewModel.isSubmitButtonActive}"
```
Why are we changing this to clickable?

Akshat-gour replied:
I had changed it back

FareesHussain replied:
This can be resolved now.

FareesHussain reviewed (changes requested):
I guess this might be causing the test failure here. PTAL

Akshat-gour committed:
changed clickable property to enable property

Akshat-gour commented:
@anandwana001 
@FareesHussain  PTAL

anandwana001 commented:
@Akshat-gour Ask for a review once all the above comments are handled or replied to by the PR author.

Akshat-gour committed:
Merge remote-tracking branch 'upstream/develop' into shifted-helper-function-to-central-testing-utility

Akshat-gour commented:
> @Akshat-gour Ask for a review once all the above comments are handled or replied to by the PR author.

i didn't get what it means

Akshat-gour commented:
>@Akshat-gour Ask for a review once all the above comments are handled or replied to by the PR author.
i didn't get what it means

@anandwana001 
@FareesHussain 
I Made the changes requested
PTAL

anandwana001 commented:
@Akshat-gour Could you please reply to the above comments in conversations?
https://github.com/oppia/oppia-android/pull/2721#discussion_r581317175
https://github.com/oppia/oppia-android/pull/2721#discussion_r581458858
https://github.com/oppia/oppia-android/pull/2721#discussion_r581458762

Akshat-gour commented:
@anandwana001 
PTAL
I replied to the comments

anandwana001 commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,47 @@
+ * Action for accessing HasErrorTest and HasNoErrorTest Functions.
+ *
+ * This is needed because Robolectric doesn't seem to properly input digits for text views using
+ * 'android:digits' or other filters. See https://github.com/robolectric/robolectric/issues/5110
```
This doesn't sound correct to me, because `TextInputAction` is all about testing `TextInputLayout` and `TextInputEditText`. Nothing is related to `android:digits`. 

> Action for accessing HasErrorTest and HasNoErrorTest Functions.

This gives a sense of the function names inside the class but not when to use this class.

Akshat-gour replied:
@anandwana001 
This class is used in testing of TextInputLayout and TextInputEditText.

This is needed Because function used at varies places so can now be defined at 
one places and can be accessed through this class object.

**Will it be good**

anandwana001 replied:
```kotlin
/**
  * Action for matching error text and performing other actions on TextInptLayout in a test infrastructure-specific way.
  */
```

anandwana001 replied:
Use this above class

Akshat-gour replied:
ok has added it

anandwana001 commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,47 @@
+  val context: Context
+) {
+  // TODO(#1720): Move this to a companion object & use a test-only singleton injector to retrieve
+  fun hasErrorText(@StringRes expectedErrorTextId: Int): Matcher<View> {
```
adding kdoc for these functions as well. 
use this reference - https://developer.android.com/kotlin/style-guide#block_tags

Akshat-gour replied:
Function hasErrorTest which is used
in org.oppia.android.app.settings.profile and in org.oppia.android.app.profile
So instead of defining function in all the place can be accessed from here.
**For Function Describtion**

anandwana001 replied:
```kotlin
/**
   * @param expectedErrorTextId is the string id which needs to match.
   * 
   * @returns a [Matcher] that matches the specified string [expectedErrorTextId] with the view's error text.
   */
```

Akshat-gour replied:
ok added it

FareesHussain reviewed (approved):
LGTM, thanks.

Akshat-gour committed:
Merge remote-tracking branch 'upstream/develop' into shifted-helper-function-to-central-testing-utility

anandwana001 commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,47 @@
+    }
+  }
+
+  fun hasNoErrorText(): Matcher<View> {
```
```kotlin
/**
   * @returns a [Matcher] that matches the view has no error text available.
   */
```

Akshat-gour replied:
ok added it

anandwana001 commented:
@Akshat-gour Please use the above-mentioned comment for kdoc.

Akshat-gour committed:
Added kdoc for the functions and class

Akshat-gour committed:
Merge remote-tracking branch 'upstream/develop' into shifted-helper-function-to-central-testing-utility

anandwana001 reviewed (approved):
LGTM, Thanks @Akshat-gour

rt4914 commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,51 @@
+ */
+class TextInputAction @Inject constructor(
+  val context: Context
+) {
```
This can be merged to single line.
`class TextInputAction @Inject constructor(val context: Context) {`

Akshat-gour replied:
done

rt4914 commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,51 @@
+class TextInputAction @Inject constructor(
+  val context: Context
+) {
+  // TODO(#1720): Move this to a companion object & use a test-only singleton injector to retrieve
```
Add `.` at the end and a blank line below this.

Akshat-gour replied:
done

rt4914 reviewed (approved):
LGTM, suggest nit changes.

Akshat-gour committed:
Did code Formatting

Akshat-gour committed:
Merge remote-tracking branch 'upstream/develop' into shifted-helper-function-to-central-testing-utility

BenHenning commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,50 @@
+  // TODO(#1720): Move this to a companion object & use a test-only singleton injector to retrieve.
+
+  /**
+   * @param expectedErrorTextId is the string id which needs to match.
+   *
+   * @returns a [Matcher] that matches the specified string [expectedErrorTextId] with the view's error text.
```
KDocs should always begin with a summary fragment before listing parameters/return value. However, given the simple contract we can actually fold both into just a summary fragment, e.g.:

```kotlin
  /**
   * Returns a [Matcher] that matches the string corresponding to the specified
   * [expectedErrorTextResId] with the view's error text.
   */
```

Ditto for the KDoc of the other method below.

Akshat-gour replied:
Did the required changes

BenHenning commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,50 @@
+   *
+   * @returns a [Matcher] that matches the specified string [expectedErrorTextId] with the view's error text.
+   */
+  fun hasErrorText(@StringRes expectedErrorTextId: Int): Matcher<View> {
```
Alluded to in my comment above, but I suggest naming the parameter ``expectedErrorTextResId`` to emphasize this is a resource ID.

Akshat-gour replied:
Did the asked changes

BenHenning commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,50 @@
+   * @returns a [Matcher] that matches the specified string [expectedErrorTextId] with the view's error text.
+   */
+  fun hasErrorText(@StringRes expectedErrorTextId: Int): Matcher<View> {
+    return object : TypeSafeMatcher<View>() {
```
Shouldn't the type being passed in be ``TextInputLayout`` so that we're properly matching against that type?

Ditto for the other method below.

Akshat-gour replied:
@BenHenning 
Its How the Function is used.It will cause to change many things to use as the function parameter

![image](https://user-images.githubusercontent.com/60266468/110094208-9496c100-7dc1-11eb-8910-6493eefc4f7d.png)

Akshat-gour replied:
@BenHenning PTAL

Akshat-gour replied:
@BenHenning 
**this is how we use the function to match.**
**Should I change to pass as parameter.**
![image](https://user-images.githubusercontent.com/60266468/110760054-4d527980-8274-11eb-81c6-15324900883e.png)

BenHenning replied:
Sorry, what I meant was: could we instead use ``TypeSafeMatcher<TextInputLayout>`` instead of ``TypeSafeMatcher<View>``? I think the former will allow us to simplify the code here since we wouldn't then need to cast the view to a ``TextInputLayout`` (``TypeSafeMatcher`` does that for us).

Akshat-gour replied:
@BenHenning 
Do you mean to use like this
![image](https://user-images.githubusercontent.com/60266468/111362205-8778ac80-86b4-11eb-9bfd-7606d02be5ae.png)

BenHenning commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Action for matching error text and performing other actions on TextInputLayout in a test infrastructure-specific way.
+ */
+class TextInputAction @Inject constructor(val context: Context) {
```
We should introduce dedicated tests for this action. That will involve creating a test-only activity with a content view that has a TextInputLayout in it that we can manipulate to verify different behaviors on this action. It's important to make sure the test utility itself works when expected since a failure in the utility can cause tests that depend on the utility to incorrectly pass, and that can result in failures being checked in & reaching our end users.

Akshat-gour replied:
How we write a test. Any help on this

BenHenning reviewed (changes requested):
Thanks @Akshat-gour! Overall I think this looks quite good. I just had a few comments--PTAL.

Akshat-gour committed:
improved kdoc

BenHenning commented on testing/src/main/java/org/oppia/android/testing/TextInputAction.kt:
```diff
@@ -34,7 +33,7 @@ class TextInputAction @Inject constructor(val context: Context) {
 
   /**
-   * @returns a [Matcher] that matches the view has no error text available.
+   * returns a [Matcher] that matches the view has no error text available.
```
```suggestion
   * Returns a [Matcher] that matches the view has no error text available.
```

Note capitalization for consistency.

Akshat-gour replied:
did the asked changes

BenHenning reviewed (changes requested):
Thanks @Akshat-gour! Left a few more comments. Please also note the comment here: https://github.com/oppia/oppia-android/pull/2721#discussion_r588008083. I don't think that has been addressed yet.

Also, @FareesHussain are you sure this PR was ready for another review? @Akshat-gour if you're ready for a follow-up review, you should at-mention the reviewers who need to take a look again so that we know to take a look. :) Also, please try to reply to all of the comments before sending a PR back for review.

BenHenning commented:
Hmm I think I reversed the assignments--sorry about that.

FareesHussain reviewed (changes requested):
@Akshat-gour Please resolve the conflicts
Reference - https://github.com/oppia/oppia-android/wiki/Android-Studio-UI-based-Github-workflow#merge-and-resolve-conflict

FareesHussain commented:
@Akshat-gour Any updates on this?

Akshat-gour commented:
@FareesHussain 
there is only one change need to be done [comment](https://github.com/oppia/oppia-android/pull/2721#discussion_r595443600)
and Change the way match happens
![image](https://user-images.githubusercontent.com/60266468/113251954-c9266b80-92e0-11eb-89bc-5b924b560069.png)

anandwana001 commented:
@Akshat-gour any update on this?

Akshat-gour commented:
@anandwana001 
sorry for the delay, actually was working on proposal creation will resume work on this.

anandwana001 commented:
@Akshat-gour any update on this?

anandwana001 commented:
CLosing this PR as no further update.
